sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/socket.scala

```diff
@@ -47,130 +48,141 @@ object TextSocketSource {
+    endOffset = end.orElse(
+      TextSocketOffset(getOffsetInternal.getOrElse(-1L))).asInstanceOf[TextSocketOffset]
   }
 
-  /** Stop this source. */
-  override def stop(): Unit = synchronized {
-    if (socket != null) {
-      try {
-        // Unfortunately, BufferedReader.readLine() cannot be interrupted, so the only way to
-        // stop the readThread is to close the socket.
-        socket.close()
-      } catch {
-        case e: IOException =>
-      }
-      socket = null
+  override def getStartOffset(): V2Offset = {
+    Option(startOffset).getOrElse(throw new IllegalStateException("start offset not set"))
+  }
+
+  override def getEndOffset(): V2Offset = {
+    Option(endOffset).getOrElse(throw new IllegalStateException("end offset not set"))
+  }
+
+  override def deserializeOffset(json: String): V2Offset = {
+    TextSocketOffset(json.toLong)
+  }
+
+  override def readSchema(): StructType = {
+    val includeTimestamp = options.getBoolean("includeTimestamp", false)
+    if (includeTimestamp) TextSocketSource.SCHEMA_TIMESTAMP else TextSocketSource.SCHEMA_REGULAR
+  }
+
+  override def createReadTasks(): JList[ReadTask[Row]] = {
+    val rawList = getBatchInternal(Option(startOffset.offset), Option(endOffset.offset))
+
+    assert(SparkSession.getActiveSession.isDefined)
+    val spark = SparkSession.getActiveSession.get
+    val numPartitions = spark.sparkContext.defaultParallelism
+
+    val slices = Array.fill(numPartitions)(new ListBuffer[(String, Timestamp)])
+    rawList.zipWithIndex.foreach { case (r, idx) =>
+      slices(idx % numPartitions).append(r)
     }
+
+    (0 until numPartitions).map { i =>
+      val slice = slices(i)
+      new ReadTask[Row] {
+        override def createDataReader(): DataReader[Row] = new DataReader[Row] {
+          private var currentIdx = -1
+
+          override def next(): Boolean = {
+            currentIdx += 1
+            currentIdx < slice.size
+          }
+
+          override def get(): Row = {
+            Row(slice(currentIdx)._1, slice(currentIdx)._2)
+          }
+
+          override def close(): Unit = {}
+        }
+      }
+    }.toList.asJava
   }
 
-  override def toString: String = s"TextSocketSource[host: $host, port: $port]"
+  override def commit(end: V2Offset): Unit = {
+    val newOffset = end.asInstanceOf[TextSocketOffset]
+    commitInternal(newOffset.offset)
+  }
+
+  override def toString: String = s"TextSocketMicroBatchReader[host: $host, port: $port]"
 }
 
-class TextSocketSourceProvider extends StreamSourceProvider with DataSourceRegister with Logging {
-  private def parseIncludeTimestamp(params: Map[String, String]): Boolean = {
-    Try(params.getOrElse("includeTimestamp", "false").toBoolean) match {
-      case Success(bool) => bool
+class TextSocketSourceProvider extends DataSourceV2
+  with MicroBatchReadSupport with StreamSourceProvider with DataSourceRegister with Logging {
```

Comment:
Why do we still need StreamSourceProvider?

Reply:
If I don't misunderstand @jose-torres 's intention, basically he wanted this socket source to work also in V1 code path.

Reply:
aah, i see earlier comments.

Reply:
TD and I discussed this offline. It should be fine to remove the V1 StreamSourceProvider implementation, because:

* this isn't a production-quality source, so users shouldn't need to fall back to it
* this source won't be particularly useful at exercising the V1 execution pipeline once we transition all sources to V2

Reply:
OK, I will update the patch accordingly.